python/lzreposync/src/lzreposync/deb_metadata_parser.py

```diff
@@ -0,0 +1,70 @@
+#  pylint: disable=missing-module-docstring
+
+import logging
+import sys
+
+from lzreposync.packages_parser import PackagesParser
+from lzreposync.translation_parser import TranslationParser
+from spacewalk.satellite_tools.syncLib import log, log2
+from spacewalk.server import rhnSQL
+from spacewalk.server.importlib import mpmSource
+
+
+class BadParserException(Exception):
+    def __init__(self, parser):
+        super().__init__(f"Bad Parser {parser}")
+
+
+#  pylint: disable-next=missing-class-docstring
+class DEBMetadataParser:
```

Comment:
This class only has one method, I would move that to a "free" function at the module level. Adding another layer makes the name longer but does not give us any benefit

Reply:
I wanted to add some abstraction when using OOP...
For example in function [import_packages_in_batch_from_parser()](https://github.com/uyuni-project/uyuni/blob/8a49313f35c903cb79926e6a5a3bcb7264b43612/python/lzreposync/src/lzreposync/import_utils.py#L89) (used in the tests) we're making use of this abstraction to import packages using abstract `parser` objects that implement the `parse_packages_metadata()` method.

And I'm afraid that using "free functions might add some complexity especially when passing a function as a parameter to another function instead of an object, in the above scenario.    

However I'm not sure whether it's the best way to approach things in our case.
(`pylint` also complains about using one method within a class `R0903: Too few public methods (1/2)`)

Reply:
To use `import_packages_in_batch_from_parser`, you first instantiate a parser object. The parser object is only used to then call the single public method it has. In other words, at the call time of `import_packages_in_batch_from_parser`, the caller has to decide which parser to pass (rpm or deb). Therefore the caller could also pass specific functions (e.g. `parse_rpm_metadata()` or `parse_deb_metadata()`) to `import_packages_in_batch_from_parser`, or call these generator functions themselves and pass the generator object instead.

In simplified/pseudo code:
```python
# parsing code
def parse_rpm_metadata(primary_xml, repository_url, arch_filter, filelists_xml, cache_dir):
    primary_parser = PrimaryParser(primary_xml, repository_url, arch_filter)
    filelists_parser = FilelistsParser(filelists_xml, cache_dir, arch_filter)
    filelitsts_parser.parse_filelists()
    for package in primary_parser.parse():
        ...
        yield rpm_package

# import code
def import_in_batches(packages, batch_size, channel=None, compatible_archs=None):
    total = 0
    for i, batch in enumerate(batched(packages, batch_size)):
        failed, *_ = import_package_batch(
            to_process=batch,
            compatible_archs=compatible_archs,
            batch_index=i,
            to_link=True,
            channel=channel
        )
        total += failed
    return total

# test code
def test_parse_and_import():
    packages = parse_rpm_metadata(primary, repo_url, arch_filter, filelists)
    failed = import_in_batches(packages, 20)
    assert ...
```

I like this more than passing a function because we decouple the different concerns. Import code shouldn't need to bother with getting the data to import, that's someone else's job.
That said, in Python it's not a problem to pass functions around, they are first-class objects, e.g.:
```python
>>> import random
>>> def foo():
...     print("Hello, I'm Foo")
... 
>>> def maybe(f):
...     if random.randint(0, 1) == 1:
...        f() 
...     else:
...        print("No luck. Maybe next time")
... 
>>> for _ in range(10):
...     maybe(foo)
... 
Hello, I'm Foo
No luck. Maybe next time
No luck. Maybe next time
No luck. Maybe next time
No luck. Maybe next time
Hello, I'm Foo
No luck. Maybe next time
Hello, I'm Foo
Hello, I'm Foo
No luck. Maybe next time

Reply:
Very good explanation..and the approach is much better too.
Thanks.